### PR TITLE
Explorer: Remove `PrintDepth` and implement `PrintIndent`.

### PR DESCRIPTION
**Pull request by prabhatexit0**
This PR removes `PrintDepth` from statement and declaration.
Implements `PrintIndent` for better indented formatting along with various changes to make printing of statements and declarations better.

### CODE REVIEW

File: explorer/interpreter/exec_program.cpp [hunk 1]
```diff
@@ -28,10 +28,13 @@ auto AnalyzeProgram(Nonnull<Arena*> arena, AST ast,
...
-    for (auto& declaration : ast.declarations) {
-      set_file_ctx.update_source_loc(declaration->source_loc());
+    for (size_t i = 0; i < ast.declarations.size(); ++i) {
+      set_file_ctx.update_source_loc(ast.declarations[i]->source_loc());
       if (trace_stream->is_enabled()) {
-        *trace_stream << *declaration;
+        *trace_stream << *ast.declarations[i];
+        if (i != ast.declarations.size() - 1) {
+          *trace_stream << "\n\n";
+        }
```
Comment (jonmeow) [inline explorer/interpreter/exec_program.cpp]: You can use llvm::Separator for cases like this. i.e.:

```
llvm::Separator sep("\n\n");
for (...) {
  *trace_stream << sep << declaration;
}
```

File: explorer/ast/declaration.h [hunk 2]
```diff
@@ -50,6 +50,14 @@ class Declaration : public AstNode {
   void Print(llvm::raw_ostream& out) const override;
   void PrintID(llvm::raw_ostream& out) const override;
 
+  // The depth parameter signals how many levels of nesting to print, so
+  // recursive calls can pass `depth - 1`, and when depth is 0 it prints "..."
+  // and returns. when depth is -1 that means "no depth limit".
```
Comment (jonmeow) [inline explorer/ast/declaration.h]:
```suggestion
  // and returns. When depth is -1 that means "no depth limit".
```

File: explorer/ast/statement.h [hunk 3]
```diff
@@ -29,9 +29,16 @@ class Statement : public AstNode {
...
 
-  void Print(llvm::raw_ostream& out) const override { PrintDepth(-1, out); }
-  void PrintID(llvm::raw_ostream& out) const override { PrintDepth(1, out); }
-  void PrintDepth(int depth, llvm::raw_ostream& out) const;
+  void Print(llvm::raw_ostream& out) const override { PrintDepth(-1, 0, out); }
+  void PrintID(llvm::raw_ostream& out) const override { PrintDepth(1, 0, out); }
+
+  // The depth parameter signals how many levels of nesting to print, so
+  // recursive calls can pass `depth - 1`, and when depth is 0 it prints "..."
+  // and returns. when depth is -1 that means "no depth limit".
```
Comment (jonmeow) [inline explorer/ast/statement.h]:
```suggestion
  // and returns. When depth is -1 that means "no depth limit".
```

File: explorer/ast/statement.cpp [hunk 4]
```diff
@@ -15,111 +15,136 @@ using llvm::cast;
...
     }
     case StatementKind::If: {
       const auto& if_stmt = cast<If>(*this);
-      out << "if (" << if_stmt.condition() << ")\n";
-      if_stmt.then_block().PrintDepth(depth - 1, out);
-      if (if_stmt.else_block()) {
-        out << "\nelse\n";
-        (*if_stmt.else_block())->PrintDepth(depth - 1, out);
+      out.indent(indent_num_spaces) << "if (" << if_stmt.condition() << ") ";
+      if (depth < 0 || depth > 1) {
```
Comment (jonmeow) [inline explorer/ast/statement.cpp]: On line 20, there's handling for depth == 0. So is this equivalent to "if (depth != 1)"?

In the prior code, when depth == 1, it would've called PrintDepth with depth == 0 which would've returned. Are these changes necessary? That is, before, wouldn't the code have printed `if (cond) ... else ...`?

File: explorer/ast/statement.cpp [hunk 5]
```diff
@@ -15,111 +15,136 @@ using llvm::cast;
...
+        if_stmt.then_block().PrintDepth(depth - 1, indent_num_spaces, out);
+        if (if_stmt.else_block()) {
+          out << "\n";
+          out.indent(indent_num_spaces) << "else\n";
+          (*if_stmt.else_block())
+              ->PrintDepth(depth - 1, indent_num_spaces + 2, out);
+        }
+      } else {
+        if (if_stmt.else_block()) {
+          out << "} else { ... }";
```
Comment (jonmeow) [inline explorer/ast/statement.cpp]: I'm a little confused by this, it doesn't look like you print the if block, so won't this result in:

`if (cond) } else { ... }`

File: explorer/testdata/trace/phase_source_program.carbon [hunk 6]
```diff
@@ -19,10 +19,7 @@ fn Main() -> i32 {
...
-// CHECK:STDOUT: fn Main ()-> i32 {
-// CHECK:STDOUT: {
-// CHECK:STDOUT: return 0;
-// CHECK:STDOUT: }
 // CHECK:STDOUT:
-// CHECK:STDOUT: }
-// CHECK:STDOUT: result: 0
+// CHECK:STDOUT: fn Main () -> i32 {
+// CHECK:STDOUT:   return 0;
+// CHECK:STDOUT: }result: 0
```
Comment (jonmeow) [inline explorer/testdata/trace/phase_source_program.carbon]: This looks like a newline is msising, can you please fix it?

File: explorer/ast/statement.cpp [hunk 7]
```diff
@@ -15,111 +15,136 @@ using llvm::cast;
...
       const auto& while_stmt = cast<While>(*this);
-      out << "while (" << while_stmt.condition() << ")\n";
-      while_stmt.body().PrintDepth(depth - 1, out);
+      out.indent(indent_num_spaces)
+          << "while (" << while_stmt.condition() << ")\n";
+      if (depth < 0 || depth > 1) {
+        while_stmt.body().PrintDepth(depth - 1, indent_num_spaces + 2, out);
+      } else {
+        out.indent(indent_num_spaces) << "...";
+      }
```
Comment (jonmeow) [inline explorer/ast/statement.cpp]: Here and in some other places, why do you print at a different indent based on the depth value?

File: explorer/ast/statement.cpp [hunk 8]
```diff
@@ -15,111 +15,136 @@ using llvm::cast;
...
       break;
     case StatementKind::VariableDefinition: {
       const auto& var = cast<VariableDefinition>(*this);
       if (var.is_returned()) {
-        out << "returned ";
+        out.indent(indent_num_spaces) << "returned ";
       }
-      out << "var " << var.pattern();
+      out.indent(var.is_returned() ? 0 : indent_num_spaces)
+          << "var " << var.pattern();
```
Comment (jonmeow) [inline explorer/ast/statement.cpp]: Can you avoid the multiple `out.indent` calls if you call it before the `if (var.is_returned())`?

File: explorer/ast/statement.cpp [hunk 9]
```diff
@@ -15,111 +15,136 @@ using llvm::cast;
 
 Statement::~Statement() = default;
 
-void Statement::PrintDepth(int depth, llvm::raw_ostream& out) const {
+void Statement::PrintDepth(int depth, int indent_num_spaces,
+                           llvm::raw_ostream& out) const {
   if (depth == 0) {
-    out << " ... ";
+    out.indent(indent_num_spaces) << " ... ";
```
Comment (jonmeow) [inline explorer/ast/statement.cpp]: It looks like every form of statement starts with an indent, except for Block.

Second, what do you think of calling `out.indent` once at the start of this PrintDepth function, instead of at the start of each statement? Given `Block`, something like:

```
if (kind() != Statement::Block) {
  out.indent(...);
}
```

Note though, we'll probably support actual block statements down the line (doesn't seem to parse right now, surprisingly), for scoping reasons. e.g.:

```
fn F() {
  // Block statement:
  {
     var scoped: MyType = ...;
  }
  // Code that depends on `scoped` being destroyed goes here.
}
```

And then those should be indented too.

File: explorer/ast/declaration.cpp [hunk 10]
```diff
@@ -15,104 +15,112 @@ using llvm::cast;
...
       break;
     }
     case DeclarationKind::ImplDeclaration: {
       const auto& impl_decl = cast<ImplDeclaration>(*this);
-      PrintID(out);
-      out << " {\n";
+      out.indent(indent_num_spaces) << PrintAsID(impl_decl) << " {\n";
       for (Nonnull<Declaration*> m : impl_decl.members()) {
-        out << *m;
+        m->PrintDepth(-1, indent_num_spaces + 2, out);
```
Comment (jonmeow) [inline explorer/ast/declaration.cpp]: Shouldn't cases like this be passing `depth - 1` if you're adding `depth` as a parameter?

File: explorer/ast/declaration.cpp [hunk 11]
```diff
@@ -15,104 +15,112 @@ using llvm::cast;
 Declaration::~Declaration() = default;
 
 void Declaration::Print(llvm::raw_ostream& out) const {
+  PrintDepth(-1, 0, out);
+}
+void Declaration::PrintDepth(int depth, int indent_num_spaces,
+                             llvm::raw_ostream& out) const {
+  if (depth == 0) {
+    out.indent(indent_num_spaces) << " ... ";
```
Comment (jonmeow) [inline explorer/ast/declaration.cpp]: Can this indent once at the start of the function, for all declarations, avoiding the call at the start of each declaration?

**Review by jonmeow — COMMENTED** [review]
zygoloid's out, so I'm taking a look at this.

I think the indentation changes are good, but I'm seeing some oddities with depth handling (I commented on examples). I've asked if there's a need to keep `depth`, maybe we can just get rid of it given that it's not used:
[#explorer question](https://discord.com/channels/655572317891461132/763516049710120960/1142124349386657923)

File: explorer/ast/statement.cpp (hunk 9 above)
Comment (jonmeow) [reply explorer/ast/statement.cpp]: Note on indenting blocks, in some cases like `while` it's going to be after a newline, won't it? i.e., you might want to test, but I think you'd get:

```
fn F() {
  while (...)
{
    // while body is indented, open curly is not.
  }
}
```

This might speak to always indenting blocks, and always putting the open curly on the next line (with appropriate indent)

File: explorer/interpreter/exec_program.cpp (hunk 1 above)
Comment (prabhatexit0) [reply explorer/interpreter/exec_program.cpp]: Oh okay but do you mean `llvm::ListSeparator`?

File: explorer/testdata/trace/phase_source_program.carbon (hunk 6 above)
Comment (prabhatexit0) [reply explorer/testdata/trace/phase_source_program.carbon]: Yes, will do in the next commit.
Thanks for pointing it out.

File: explorer/ast/statement.cpp (hunk 9 above)
Comment (prabhatexit0) [reply explorer/ast/statement.cpp]: Oh yes, these issues can arise with the current way of indentation. Thanks for pointing this out!
I was thinking can we have formatting like this?

```
fn Foo() -> type
{
  // body
  if(...)
  {
    ...
  }
}
```

instead of 
```
fn Foo() -> type {
  // body
  if(...) {
    ...
  }
}
```

It would be a bit simpler to implement.
What do you think?

File: explorer/ast/declaration.cpp (hunk 11 above)
Comment (prabhatexit0) [reply explorer/ast/declaration.cpp]: Sure, I'll do that

File: explorer/ast/declaration.cpp (hunk 10 above)
Comment (prabhatexit0) [reply explorer/ast/declaration.cpp]: I wasn't exactly sure how the `depth` thing was working by the time I wrote this code.
I'm waiting for your response in the **#explorer** on discord regarding if we need to keep `PrintDepth` or not before working on this change.

File: explorer/ast/statement.cpp (hunk 8 above)
Comment (prabhatexit0) [reply explorer/ast/statement.cpp]: Sure, I'll do it in the next commit.

File: explorer/ast/statement.cpp (hunk 7 above)
Comment (prabhatexit0) [reply explorer/ast/statement.cpp]: I wasn't exactly sure how the `depth` thing was working by the time I wrote this code.
I'm waiting for your response in the **#explorer** on discord regarding if we need to keep `PrintDepth` or not before working on this change.  (2)

File: explorer/ast/statement.cpp (hunk 5 above)
Comment (prabhatexit0) [reply explorer/ast/statement.cpp]: Oh, I'll fix this in the next commit.

File: explorer/ast/statement.cpp (hunk 4 above)
Comment (prabhatexit0) [reply explorer/ast/statement.cpp]: I wasn't exactly sure how the `depth` thing was working by the time I wrote this code.
I'm waiting for your response in the **#explorer** on discord regarding if we need to keep `PrintDepth` or not before working on this change. (3)

File: explorer/ast/declaration.cpp (hunk 10 above)
Comment (jonmeow) [reply explorer/ast/declaration.cpp]: Let's go ahead with that, maybe also renaming PrintDepth to PrintIndented or similar?

File: explorer/ast/statement.cpp (hunk 9 above)
Comment (jonmeow) [reply explorer/ast/statement.cpp]: I think that'd be fine, yes.

**Review by jonmeow — COMMENTED** [review]
As noted on #explorer, let's go ahead with removing depth. I'm betting against objections since it's been a couple days, and it's unused in code. :)

File: explorer/ast/declaration.cpp (hunk 10 above)
Comment (prabhatexit0) [reply explorer/ast/declaration.cpp]: Done, renamed it to `PrintIndent`.

File: explorer/ast/declaration.h [hunk 12]
```diff
@@ -238,8 +240,8 @@ class CallableDeclaration : public Declaration {
         body_(context.Clone(other.body_)),
         virt_override_(other.virt_override_) {}
 
-  void PrintDepth(int depth, llvm::raw_ostream& out) const;
-
+  void PrintIndent(int indent_num_spaces,
+                   llvm::raw_ostream& out) const override;
```
Comment (jonmeow) [inline explorer/ast/declaration.h:244]: So you know, I'm looking at this override and I think we should probably remove it (inlining the code) to reduce complexity and be more consistent (i.e., removing the indent special-casing for callables). It'd also mean PrintIndent wouldn't need to be virtual. I'm fine with it for now since the function was there before, I just want you to know the line of thought.

File: explorer/ast/statement.cpp [hunk 13]
```diff
@@ -15,39 +15,86 @@ using llvm::cast;
 
 Statement::~Statement() = default;
 
-void Statement::PrintDepth(int depth, llvm::raw_ostream& out) const {
-  if (depth == 0) {
-    out << " ... ";
-    return;
+void Statement::PrintID(llvm::raw_ostream& out) const {
```
Comment (jonmeow) [inline explorer/ast/statement.cpp:18]: While this is replacing the PrintDepth(1, ...) call, will this work as intended? Every use of `Print`, including the implicit ones like `<< match.expression() <<`, corresponds to a fully recursive call.

Maybe it should just print a keyword or brief placeholder matching the statement? e.g., `match` for Match, `{ ... }` for Block, PrintID for ExpressionStatement, etc. This would be similar to how expressions print IDs.

File: explorer/ast/statement.cpp [hunk 14]
```diff
@@ -15,39 +15,86 @@ using llvm::cast;
...
+    case StatementKind::ReturnExpression: {
+      Print(out);
+    }
+  }
+}
+
+void Statement::PrintIndent(int indent_num_spaces,
+                            llvm::raw_ostream& out) const {
+  if (kind() != StatementKind::Block) {
+    out.indent(indent_num_spaces);
```
Comment (jonmeow) [inline explorer/ast/statement.cpp]: Under the current approach, Block is also indented, so this conditional shouldn't be needed.

File: explorer/ast/statement.cpp [hunk 15]
```diff
@@ -79,16 +126,19 @@ void Statement::PrintDepth(int depth, llvm::raw_ostream& out) const {
...
+      out << (inc_dec.is_increment() ? "++" : "--") << inc_dec.argument()
+          << ";";
       break;
     }
     case StatementKind::If: {
       const auto& if_stmt = cast<If>(*this);
-      out << "if (" << if_stmt.condition() << ")\n";
-      if_stmt.then_block().PrintDepth(depth - 1, out);
+      out << "if (" << if_stmt.condition() << ")";
+      out << "\n";
```
Comment (jonmeow) [inline explorer/ast/statement.cpp]:
```suggestion
      out << "if (" << if_stmt.condition() << ")\n";
```

File: explorer/ast/statement.cpp [hunk 16]
```diff
@@ -79,16 +126,19 @@ void Statement::PrintDepth(int depth, llvm::raw_ostream& out) const {
...
+      if_stmt.then_block().PrintIndent(indent_num_spaces, out);
       if (if_stmt.else_block()) {
-        out << "\nelse\n";
-        (*if_stmt.else_block())->PrintDepth(depth - 1, out);
+        out << "\n";
+        out.indent(indent_num_spaces) << "else\n";
+        (*if_stmt.else_block())->PrintIndent(indent_num_spaces + 2, out);
```
Comment (jonmeow) [inline explorer/ast/statement.cpp]:
```suggestion
        (*if_stmt.else_block())->PrintIndent(indent_num_spaces, out);
```

File: explorer/ast/statement.cpp [hunk 17]
```diff
@@ -15,39 +15,86 @@ using llvm::cast;
...
-      while_stmt.body().PrintDepth(depth - 1, out);
+      while_stmt.body().PrintIndent(indent_num_spaces, out);
       break;
     }
     case StatementKind::For: {
       const auto& for_stmt = cast<For>(*this);
       out << "for (" << for_stmt.variable_declaration() << " in "
           << for_stmt.loop_target() << ")\n";
-      for_stmt.body().PrintDepth(depth - 1, out);
+      for_stmt.body().PrintIndent(indent_num_spaces + 2, out);
```
Comment (jonmeow) [inline explorer/ast/statement.cpp]:
```suggestion
      for_stmt.body().PrintIndent(indent_num_spaces, out);
```

With the way blocks are printed, they shouldn't have an increased indent.

File: explorer/ast/statement.cpp [hunk 18]
```diff
@@ -15,39 +15,86 @@ using llvm::cast;
...
   switch (kind()) {
     case StatementKind::Match: {
       const auto& match = cast<Match>(*this);
-      out << "match (" << match.expression() << ") {";
-      if (depth < 0 || depth > 1) {
+      out << "match (" << match.expression() << ") {\n";
+      for (const auto& clause : match.clauses()) {
+        out.indent(indent_num_spaces + 2)
+            << "case " << clause.pattern() << " => \n";
+        clause.statement().PrintIndent(indent_num_spaces + 2, out);
```
Comment (jonmeow) [inline explorer/ast/statement.cpp:97]:
```suggestion
        clause.statement().PrintIndent(indent_num_spaces, out);
```

This is actually always a block, so probably shouldn't increase indent.

For reference, from parser.ypp:
```
clause:
  CASE pattern DOUBLE_ARROW block
    { $$ = Match::Clause($[pattern], $[block]); }
| DEFAULT DOUBLE_ARROW block
    {
      $$ = Match::Clause(arena->New<BindingPattern>(
                             context.source_loc(), std::string(AnonymousName),
                             arena->New<AutoPattern>(context.source_loc()),
                             ExpressionCategory::Value),
                         $[block]);
    }
;
```

**Review by jonmeow — COMMENTED** [review]
Thanks! I think this is the right direction, though we probably still need to simplify PrintID to avoid recursive printing.

File: explorer/ast/declaration.h (hunk 12 above)
Comment (prabhatexit0) [reply explorer/ast/declaration.h:244]: Thanks for letting me know about it! I can do something about it in some other PR.

File: explorer/ast/statement.cpp (hunk 13 above)
Comment (prabhatexit0) [reply explorer/ast/statement.cpp:18]: Oh I'm sorry I missed that. I'll do that in the next commit. Thanks!

File: explorer/ast/statement.cpp (hunk 14 above)
Comment (prabhatexit0) [reply explorer/ast/statement.cpp]: Okay. I will remove the conditional in the next commit.

File: explorer/ast/statement.cpp (hunk 17 above)
Comment (prabhatexit0) [reply explorer/ast/statement.cpp]: Thanks for pointing this out!

File: explorer/ast/statement.cpp (hunk 18 above)
Comment (prabhatexit0) [reply explorer/ast/statement.cpp:97]: Doing this is leading to this output"
```
    match (self.element) {
      case OptionalElement(T).None =>
    {
      return false;
    }
    }
```

Note: for clause the indent is `indent_num_spaces + 2` as it is nested inside the match block.

File: explorer/ast/statement.cpp (hunk 13 above)
Comment (prabhatexit0) [reply explorer/ast/statement.cpp:18]: After trying it out I found out that printing some elements with very less detail was making it less readable in the trace output. 
So, for block I'm doing `PrintAsID` for the statements it contains, also doing `PrintAsID` for few other cases.
What do you think about this?

Comment (jonmeow) [reply explorer/ast/statement.cpp:18]: I think that PrintID should not do anything that is recursive or loops. If you are fine with recursion or looping, you should probably be using Print instead. [note, part of the issue with looping in a block is that you don't know how many statements there are, and there may be lots]

File: explorer/ast/statement.cpp (hunk 18 above)
Comment (jonmeow) [reply explorer/ast/statement.cpp:97]: Ah, thanks, I'd missed it matching the clause indent.

File: explorer/ast/statement.cpp (hunk 13 above)
Comment (prabhatexit0) [reply explorer/ast/statement.cpp:18]: Oh, to get rid of this looping issue with the block can we have the following cases?

if the number of statements inside the block are less than or equal to 2
```
{id stmt 1; id stmt 2}
```
if the statements in the block are greater than 2
```
{id stmt 1; ... id stmt2}
```

although in the trace output, source location is also added in the cases when `PrintID` is used but printing `{...}` for blocks is making it a bit hard to understand which part of code it belongs to.

Comment (jonmeow) [reply explorer/ast/statement.cpp:18]: That would be fine, but along similar lines you might consider:

0: `{}`
1: `{id stmt 1;}`
N: `{id stmt 1; ...[N-1] statements...}`

i.e., after the 1st statement, just print the # of statements that were elided.

Comment (prabhatexit0) [reply explorer/ast/statement.cpp:18]: Nice suggestion, I will implement that.

Comment (prabhatexit0) [reply explorer/ast/statement.cpp:18]: I also used `PrintID` at places where I felt like the information could be useful. Let me know what you think about it.
And I noticed an issue in expression's `PrintID` due to it there was an assertion error happening, I also tried to fix that.

Comment (jonmeow) [reply explorer/ast/statement.cpp:18]: Thanks for pointing out the changes. I'm still concerned about this: while a Block statement would just be `{ ... }` which could be confusing, there are more cues (keywords) for other pieces. I think too that not recursing is more in keeping with the prior PrintDepth implementation.

Note especially that your changes to For are back to recursively printing the entire for loop contents. This is returning to the issue I was trying to point out originally, that PrintID needs to not be fully recursive.

I would appreciate going back to a simple implementation so that we can finish review.

Comment (prabhatexit0) [reply explorer/ast/statement.cpp:18]: > I'm still concerned about this: while a Block statement would just be { ... } which could be confusing, there are more cues (keywords) for other pieces. I think too that not recursing is more in keeping with the prior PrintDepth implementation.

Oh okay. Thanks for letting me know about that!

> Note especially that your changes to For are back to recursively printing the entire for loop contents. This is returning to the issue I was trying to point out originally, that PrintID needs to not be fully recursive.

I actually forgot to use `PrintAsID(...)` there but going back to simple implementation.

> I would appreciate going back to a simple implementation so that we can finish review.

I'm reverting back to the simple implementation in next commit.

Comment (jonmeow) [reply explorer/ast/statement.cpp:18]: I do understand that non-recursive yields less information, but I think things like `{var x: T = {}.(interface ImplicitAs(T = class T).Convert)();return 0;}` (which had been in trace output before) are actually more indicative of an issue that needed to be fixed, that arbitrary amounts of output were getting printed.

I think this current setup is good, thanks!

**Review by jonmeow — APPROVED** [review]
Thanks for all the changes! I'll go ahead and merge this.

**Comment by prabhatexit0** [comment]
> Thanks for all the changes! I'll go ahead and merge this.

Thanks for the review!
